src/frontend/cellprofiler/modules/correctilluminationapply.py

```diff
@@ -248,50 +245,33 @@ def run_image(self, image, workspace):
-                    :, :, numpy.newaxis
-                ]
-        # Throw an error if image and illum data are incompatible
-        if orig_image.pixel_data.shape[:2] != illum_function_pixel_data.shape[:2]:
-            raise ValueError(
-                "This module requires that the image and illumination function have equal dimensions.\n"
-                "The %s image and %s illumination function do not (%s vs %s).\n"
-                "If they are paired correctly you may want to use the Resize or Crop module to make them the same size."
-                % (
-                    image_name,
-                    illum_correct_name,
-                    orig_image.pixel_data.shape,
-                    illum_function_pixel_data.shape,
-                )
-            )
         #
-        # Either divide or subtract the illumination image from the original
+        # Apply the illumination function
         #
-        if image.divide_or_subtract == DOS_DIVIDE:
-            output_pixels = orig_image.pixel_data / illum_function_pixel_data
-        elif image.divide_or_subtract == DOS_SUBTRACT:
-            output_pixels = orig_image.pixel_data - illum_function_pixel_data
-            output_pixels[output_pixels < 0] = 0
-        else:
-            raise ValueError(
-                "Unhandled option for divide or subtract: %s"
-                % image.divide_or_subtract.value
+        try:
+            output_pixels = correct_illumination_apply(
+                orig_image.pixel_data,
+                illum_function_pixel_data,
+                image.divide_or_subtract.value,
+                truncate_low=self.truncate_low.value,
+                truncate_high=self.truncate_high.value,
             )
-        
-        #
-        # Optionally, clip high and low values
-        #
-        if self.truncate_low.value:
-            output_pixels = numpy.where(output_pixels < 0, 0, output_pixels)
-        if self.truncate_high.value:
-            output_pixels = numpy.where(output_pixels > 1, 1, output_pixels)
-        
+        except ValueError as e:
+            # Update error message to be more descriptive if the error is due to image and illum function being different shapes
+            if "This module requires that the image and illumination function have equal dimensions" in str(e):
+                raise ValueError(
+                    "This module requires that the image and illumination function have equal dimensions.\n"
+                    "The %s image and %s illumination function do not (%s vs %s).\n"
+                    "If they are paired correctly you may want to use the Resize or Crop module to make them the same size."
+                    % (
+                        image_name,
+                        illum_correct_name,
+                        orig_image.pixel_data.shape,
+                        illum_function_pixel_data.shape,
+                    )
+                )
+            else:
+                raise e
```

Comment:
Keeping the validation (and auto-fix) of illum function in the FE. Fixing an invalid illum function is not the library's responsibility and would be unfair to other modules that do not auto-fix invalid inputs. Instead, I've added an assert in the library to validate that the shapes match and throw a message otherwise.